server.go

```diff
@@ -982,9 +984,32 @@ func (s *server) sendPeerMessages(
 		defer wg.Done()
 	}
 
+	// We queue each message, creating a slice of error channels that
+	// can be inspected after every message is successfully added to
+	// the queue.
+	var errChans []chan error
 	for _, msg := range msgs {
-		targetPeer.queueMsg(msg, nil)
+		errChan := make(chan error, 1)
+		if err := targetPeer.queueMsg(msg, errChan); err != nil {
+			return err
+		}
+		errChans = append(errChans, errChan)
 	}
+
+	// Return an error if any of the messages failed being sent.
```

Comment:
Nice job here as well! This has much better interleaving properties :) 

For both sends and broadcasts, this will still hold the server's mutex until all messages have been ACKd or failures occurs. With a few small changes, we can make this only block during queueing, and allow the receiver to conditionally handle errors outside the mutex. This should help keep large sends or broadcasts from limiting our throughput and minimize contention in the server.

1. Have `sendPeerMessages` and `sendToPeer` return `([]chan error, error)`
2. Remove the remainder of this section into a separate method such as `s.handleErrChans([]chan error) error`
3. Modify `SendToPeer` to handle error chans, such as:
```golang
s.Lock()
errChans, err := s.sendToPeer(target, msgs)
s.Unlock()
if err != nil {
    return err
}

return s.handleErrChans(errChans)
```

We won't need to modify broadcasting at all because we don't care about the errors and this actually returns it to the desired send-and-forget behavior. I suspect the compiler might even be able to optimize out the process of appending errChans during broadcasts, since the return variable can't be captured when run as a go routine.

Reply:
If the compiler optimization doesn't happen, we can always shave the extra memory allocations by using the nil-ness of the wait group to instruct whether or not to instantiate error chans and appending to the slice.

Reply:
Good idea. If doing it this way, I would argue that we might just make the methods return _only_ `[]chan error`. Since a broadcast does not distinguish between failing to add a message to the queue, and failure sending at a later point, and a "reliable send" will also treat these two error scenarios similarly, we might as well only check the error channels if we want to.

The only downside I see by doing this is that if we send a batch of messages, and the first one cannot be queued, we will continue trying to queue the rest regardless. I don't think this will usually be a problem (they will all just fail quickly), but in the unlikely scenario that one succeeds (while a previous one failed) we end up with messages out of order. Not even sure this is possible 😛 

We _can_ make `sendPeerMessages` return only `[]chan error` as mentioned above, but still keep the return error from `queueMsg`. In the case of failing to queue a message, we can send this error directly on the remaining error chans, avoiding the problem I mentioned above.

Thoughts?